readthedocs/doc_builder/builder.py

```diff
@@ -0,0 +1,443 @@
+import os
+from collections import defaultdict
+
+import structlog
+from django.conf import settings
+from django.utils.translation import gettext_lazy as _
+
+from readthedocs.builds.constants import EXTERNAL
+from readthedocs.doc_builder.config import load_yaml_config
+from readthedocs.doc_builder.loader import get_builder_class
+from readthedocs.doc_builder.python_environments import Conda, Virtualenv
+from readthedocs.projects.exceptions import RepositoryError
+from readthedocs.projects.models import Feature
+from readthedocs.projects.signals import after_build, before_build, before_vcs
+
+log = structlog.get_logger(__name__)
+
+
+class DocumentationBuilder:
```

Comment:
Either here, or at the top-level of this file would be a great place to explain what this code is responsible for. There's some of this in the PR description, but would be helpful in the code here too. Is this any filesystem-related activity, or just the entire process of whatever executing an "RTD build" does?

Reply:
This class is in charge of doing _all the steps required to build the documentation_ (running all the commands needed) but won't do anything related to readthedocs itself (e.g update build's status, sync versions, trigger any task, create search index, etc)

I can add more docstrings, yeah.